### PR TITLE
Stream output from `pull` and `push` command

### LINKED ISSUE #677: Stream log lines when doing `pull`/`push`
Hi! Thank you for the great library. We used it for pipelines that involve building and pushing a lot of images to private repositories. While the `build` interface does support `stream_logs`, would it be possible to also add something like this to `pull`/`push`?

### PR DESCRIPTION
**Pull request by dabico**
Support streaming output from the `pull` and `push` commands.

## Pulling

Minimal reproducible example of pulling multiple images with log streaming:

```python
import logging
from logging import INFO, Formatter, StreamHandler
from sys import stdout

from python_on_whales import docker

formatter = Formatter("%(asctime)s [%(levelname)s] %(message)s")
handler = StreamHandler(stream=stdout)
logger = logging.getLogger(__file__)
logger.setLevel(INFO)
handler.setFormatter(formatter)
logger.addHandler(handler)

images = [
    "registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1",
    "registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2",
]

if __name__ == '__main__':
    if generator := docker.image.pull(images, stream_logs=True):
        for image, line in generator:
            logger.info("[%s] %s", image, line.decode(errors="replace").strip())
```

Produces the following output:

```
2025-06-07 12:53:54,627 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 1: Pulling from p/ml-4-se-lab/python-on-whales/test
2025-06-07 12:53:54,661 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 2: Pulling from p/ml-4-se-lab/python-on-whales/test
2025-06-07 12:53:55,693 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 771aa65b5adb: Pulling fs layer
2025-06-07 12:53:55,693 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] d69d4d41cfe2: Pulling fs layer
2025-06-07 12:53:55,703 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 771aa65b5adb: Pulling fs layer
2025-06-07 12:53:55,703 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] d69d4d41cfe2: Pulling fs layer
2025-06-07 12:53:57,139 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 771aa65b5adb: Download complete
2025-06-07 12:53:57,139 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] d69d4d41cfe2: Download complete
2025-06-07 12:53:57,150 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 771aa65b5adb: Download complete
2025-06-07 12:53:57,150 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] d69d4d41cfe2: Download complete
2025-06-07 12:53:57,171 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 771aa65b5adb: Pull complete
2025-06-07 12:53:57,173 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 771aa65b5adb: Pull complete
2025-06-07 12:53:57,174 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] d69d4d41cfe2: Pull complete
2025-06-07 12:53:57,174 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] Digest: sha256:484d7d8f6639713a572ded4562fd1106acb69424b7d218bac0944406c279396a
2025-06-07 12:53:57,174 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] Status: Downloaded newer image for registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1
2025-06-07 12:53:57,176 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1
2025-06-07 12:53:57,179 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] d69d4d41cfe2: Pull complete
2025-06-07 12:53:57,179 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] Digest: sha256:c02bb7a2b95fdffc7e0dd75157260ed38df4a81a8cc1e9412a42441ebfb28e9d
2025-06-07 12:53:57,179 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] Status: Downloaded newer image for registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2
2025-06-07 12:53:57,180 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2
```

## Pushing

Minimal reproducible example of pushing multiple images with log streaming:

```python
import logging
from logging import INFO, Formatter, StreamHandler
from sys import stdout

from python_on_whales import docker

formatter = Formatter("%(asctime)s [%(levelname)s] %(message)s")
handler = StreamHandler(stream=stdout)
logger = logging.getLogger(__file__)
logger.setLevel(INFO)
handler.setFormatter(formatter)
logger.addHandler(handler)

images = [
    "registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1",
    "registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2",
]

if __name__ == '__main__':
    if generator := docker.image.push(images, stream_logs=True):
        for image, line in generator:
            logger.info("[%s] %s", image, line.decode(errors="replace").strip())
```

Produces the following output:

```
2025-06-01 19:26:35,408 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] The push refers to repository [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test]
2025-06-01 19:26:35,418 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] The push refers to repository [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test]
2025-06-01 19:26:35,515 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 667acca900bd: Waiting
2025-06-01 19:26:35,515 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 771aa65b5adb: Waiting
2025-06-01 19:26:35,515 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] d69d4d41cfe2: Waiting
2025-06-01 19:26:35,521 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 0428fa5aaf34: Waiting
2025-06-01 19:26:35,521 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 771aa65b5adb: Waiting
2025-06-01 19:26:35,521 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] d69d4d41cfe2: Waiting
2025-06-01 19:26:35,613 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 771aa65b5adb: Waiting
2025-06-01 19:26:35,613 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] d69d4d41cfe2: Waiting
2025-06-01 19:26:35,613 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 667acca900bd: Waiting
2025-06-01 19:26:35,619 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 0428fa5aaf34: Waiting
2025-06-01 19:26:35,620 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 771aa65b5adb: Waiting
2025-06-01 19:26:35,620 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] d69d4d41cfe2: Waiting
2025-06-01 19:26:35,712 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 667acca900bd: Waiting
2025-06-01 19:26:35,712 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 771aa65b5adb: Waiting
2025-06-01 19:26:35,712 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] d69d4d41cfe2: Waiting
2025-06-01 19:26:35,720 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 0428fa5aaf34: Waiting
2025-06-01 19:26:35,720 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 771aa65b5adb: Waiting
2025-06-01 19:26:35,720 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] d69d4d41cfe2: Waiting
2025-06-01 19:26:35,816 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 667acca900bd: Waiting
2025-06-01 19:26:35,816 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 771aa65b5adb: Waiting
2025-06-01 19:26:35,816 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] d69d4d41cfe2: Waiting
2025-06-01 19:26:35,820 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 0428fa5aaf34: Waiting
2025-06-01 19:26:35,820 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 771aa65b5adb: Waiting
2025-06-01 19:26:35,820 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] d69d4d41cfe2: Waiting
2025-06-01 19:26:35,911 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 771aa65b5adb: Layer already exists
2025-06-01 19:26:35,911 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] d69d4d41cfe2: Waiting
2025-06-01 19:26:35,911 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 667acca900bd: Waiting
2025-06-01 19:26:35,921 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 0428fa5aaf34: Waiting
2025-06-01 19:26:35,921 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 771aa65b5adb: Waiting
2025-06-01 19:26:35,921 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] d69d4d41cfe2: Layer already exists
2025-06-01 19:26:36,010 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 667acca900bd: Waiting
2025-06-01 19:26:36,010 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] d69d4d41cfe2: Waiting
2025-06-01 19:26:36,020 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 0428fa5aaf34: Already exists
2025-06-01 19:26:36,020 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 771aa65b5adb: Layer already exists
2025-06-01 19:26:36,111 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 667acca900bd: Already exists
2025-06-01 19:26:36,111 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] d69d4d41cfe2: Layer already exists
2025-06-01 19:26:37,106 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:1] 1: digest: sha256:484d7d8f6639713a572ded4562fd1106acb69424b7d218bac0944406c279396a size: 855
2025-06-01 19:26:37,109 [INFO] [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:2] 2: digest: sha256:c02bb7a2b95fdffc7e0dd75157260ed38df4a81a8cc1e9412a42441ebfb28e9d size: 855
```

## Notes

I've also updated both methods to filter duplicates from the image list while preserving the order in which they were specified.

Fixes #677

### CODE REVIEW

**Comment by dabico** [comment]
Naive implementation currently only streams output for only one image pushed. A script like:

```python
from python_on_whales import docker

images = [
    "registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test:latest",
]
for line in docker.image.push(images) or ():
    print(line.strip())

```

Will produce the following output:

```
The push refers to repository [registry.jetbrains.team/p/ml-4-se-lab/python-on-whales/test]
f52001c07a1d: Waiting
79da15229fa6: Waiting
6e771e15690e: Waiting
f52001c07a1d: Waiting
79da15229fa6: Waiting
6e771e15690e: Waiting
f52001c07a1d: Waiting
79da15229fa6: Waiting
6e771e15690e: Waiting
f52001c07a1d: Waiting
79da15229fa6: Waiting
6e771e15690e: Waiting
f52001c07a1d: Waiting
79da15229fa6: Waiting
6e771e15690e: Waiting
79da15229fa6: Waiting
6e771e15690e: Waiting
f52001c07a1d: Waiting
79da15229fa6: Waiting
6e771e15690e: Waiting
f52001c07a1d: Waiting
79da15229fa6: Waiting
6e771e15690e: Waiting
f52001c07a1d: Waiting
f52001c07a1d: Waiting
79da15229fa6: Waiting
6e771e15690e: Waiting
79da15229fa6: Pushed
f52001c07a1d: Pushed
6e771e15690e: Pushed
latest: digest: sha256:584f3c602adbc8abed8443c563ee775eb2b1192291b9b74fb85575b857347be7 size: 855
```

Need to investigate failing tests and add my own.

**Comment by dabico** [comment]
Hi! I'm still exploring your codebase and working on the feature whenever I find the time, hence the ongoing "draft" state. I guess my only question right now would be: Should I introduce a parameter for output streaming to mirror the implementation of `build`? As it stands, the method immediately returns the output generator before the `push` is finished. Not sure if this is desirable. For instance, I had to modify the tests to wait for the push to complete before proceeding with operations that depend on the pushed image.

**Comment by dabico** [comment]
Hey @gabrieldemarmiesse, it's ready now ;)

**Comment by dabico** [comment]
I've added the same capability to `pull`. If you're still reviewing the code and have some suggestions don't worry: most of what I've added is equivalent to what I did in `push`. As such any comments pointed to one will by extension apply for the other.

**Comment by gabrieldemarmiesse** [comment]
Thanks, I'm on holidays, but I should be able to review it in the next few days, apologies for the delay!

**Review by gabrieldemarmiesse — CHANGES_REQUESTED** [review]
Nice the implementation looks good as far as I can tell. Could you add unit tests for the logs streaming? This should be fairly straightforward if you look at the other unit tests of Python-on-whales. Since the code is fairly complexe, I'd be happy if you could be extensive and try as many corner cases as you can. Very cool PR!

**Comment by dabico** [comment]
Hi Gabriel! Been busy this week, so I had little time to do this. Up to this point I only did very basic manual testing. I started writing more tests now and discovered a bug in which exceptions are not properly raised when streaming. Will look into it and let you know once it's resolved.

**Comment by gabrieldemarmiesse** [comment]
Take your time! Aand if you want to hindle only the single-threaded case, it's fine too :)

**Comment by dabico** [comment]
I added some tests, but they are too specific to `docker`. Need to mess around with `podman` for a bit and see how I can generify the failing tests.

**Comment by dabico** [comment]
Evening @gabrieldemarmiesse! I added some tests, let me know if you'd want more. I struggled quite a bit, since tests I wrote would pass locally, but not in CI.

File: python_on_whales/components/image/cli_wrapper.py [hunk 1]
```diff
@@ -503,49 +506,133 @@ def pull(
...
-            return [self._pull_single_tag(x[0], quiet=quiet, platform=platform)]
+            result = self._pull_single_tag(
+                image_name=x,
+                quiet=quiet,
+                stream_logs=stream_logs,
+                platform=platform,
+            )
+            return result if stream_logs else Image(self.client_config, x)
+
+        images = list(dict.fromkeys(x))
```
Comment (gabrieldemarmiesse) [inline python_on_whales/components/image/cli_wrapper.py:545]: I'm not exactly sure why this is needed. Isn't 
```suggestion
        images = list(x)
```
fine too?

File: python_on_whales/components/image/cli_wrapper.py [hunk 2]
```diff
@@ -503,49 +506,133 @@ def pull(
...
+            x: Image name(s), can be a string or a list of strings.
+                In case a list is passed, multithreading is used to pull
+                the images.
             quiet: If you don't want to see the progress bars.
+            stream_logs: If `False` this function returns the pulled image(s).
+                If `True`, this function returns an `Iterable` of `Tuple[str, bytes]`
+                where the first element corresponds to the image name that a pull is being done for.
+                The second element is the log statement related to pull progress, as `bytes`.
+                You'll need to call `.decode()` if you want the logs as `str`.
+                See [the streaming guide](https://gabrieldemarmiesse.github.io/python-on-whales/user_guide/docker_run/#stream-the-output)
```
Comment (gabrieldemarmiesse) [inline python_on_whales/components/image/cli_wrapper.py:526]: The streaming guide is not very helpful (and can be even confusing here as it's not the same format). You can just remove this line. Same for push.

**Review by gabrieldemarmiesse — COMMENTED** [review]
Very nice PR! Thanks for all this hard work!

**Review by gabrieldemarmiesse — APPROVED** [review]

File: python_on_whales/components/image/cli_wrapper.py (hunk 1 above)
Comment (gabrieldemarmiesse) [reply python_on_whales/components/image/cli_wrapper.py]: Thanks for the explanation. Indeed that makes sense. I'll add it back, my bad
